tools/ts_source_copy_check.py

```diff
@@ -0,0 +1,205 @@
+    filename = os.path.join(rootdir, file_to_format.filename)
+
+    # Extract the language code from the filename
+    # (assuming format: mixxx_<lang_code>.ts)
+    match = re.search(r".*/mixxx_(\w+)\.ts", file_to_format.filename)
+    if not match:
+        return False
+
+    lang_code = match.group(1)
+
+    if lang_code.startswith("en_"):
+        # for Engish languages all source copies are allowed
+        return False
+
+    parser = etree.XMLParser(recover=True)
+
+    try:
+        tree = etree.parse(filename, parser)
+    except Exception as e:
+        print("XML parsing failed:")
+        print(e)
+        return False
+
+    root = tree.getroot()
+
+    found = False
+
+    for context in root.findall("context"):
+        for message in context.findall("message"):
+            source = message.find("source").text
+            translation_elem = message.find("translation")
+            if translation_elem is None:
+                continue
+            translation = translation_elem.text
+            if source != translation:
+                continue
+            line_number = translation_elem.sourceline
+            for start, end in file_to_format.lines:
+                if line_number >= start and line_number <= end:
+                    if is_untranstaled_allowed(source, lang_code):
+                        continue
+                    print(
+                        "Source copy found at: "
+                        f"{file_to_format.filename}:{line_number}\n{source}",
+                        file=sys.stderr,
+                    )
+                    add_to_allow_list(source, lang_code)
+                    found = True
+    return found
+
+
+def main(argv: typing.Optional[typing.List[str]] = None) -> int:
+    logging.basicConfig(
+        format="[%(levelname)s] %(message)s", level=logging.INFO
+    )
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--from-ref", help="use changes changes since commit")
+    parser.add_argument("--to-ref", help="use changes until commit")
+    parser.add_argument("files", nargs="*", help="only check these files")
+    args = parser.parse_args(argv)
+
+    if not args.from_ref:
+        args.from_ref = os.getenv("PRE_COMMIT_FROM_REF") or os.getenv(
+            "PRE_COMMIT_SOURCE"
+        )
+
+    if not args.to_ref:
+        args.to_ref = os.getenv("PRE_COMMIT_TO_REF") or os.getenv(
+            "PRE_COMMIT_ORIGIN"
+        )
+
+    # Filter filenames
+    rootdir = githelper.get_toplevel_path()
+
+    files_with_added_lines = githelper.get_changed_lines_grouped(
+        from_ref=args.from_ref,
+        to_ref=args.to_ref,
+        filter_lines=lambda line: line.added,
+        include_files=args.files,
+    )
+
+    ret = 0
+    for changed_file in files_with_added_lines:
+        if check_copied_source_on_lines(rootdir, changed_file):
+            ret = 1
+
+    if ret:
+        print(
+            "\n"
```

Comment:
```suggestion
```

Reply:
This is desired to have a distance to the individual complains.